src/rfc3161_client/verify.py

```diff
@@ -12,170 +10,328 @@
 from rfc3161_client.tsp import PKIStatus, TimeStampRequest, TimeStampResponse
 
 
-@dataclass
-class VerifyOpts:
-    policy_id: cryptography.x509.ObjectIdentifier | None
-    tsa_certificate: cryptography.x509.Certificate | None
-    intermediates: list[cryptography.x509.Certificate]
-    roots: list[cryptography.x509.Certificate]
-    nonce: int
-    common_name: str
-
+class VerifyBuilder:
+    def __init__(
+        self,
+        policy_id: cryptography.x509.ObjectIdentifier | None = None,
+        tsa_certificate: cryptography.x509.Certificate | None = None,
+        intermediates: list[cryptography.x509.Certificate] | None = None,
+        roots: list[cryptography.x509.Certificate] | None = None,
+        nonce: int | None = None,
+        common_name: str | None = None,
+    ):
+        """Init method."""
+        self._policy_id: cryptography.x509.ObjectIdentifier | None = policy_id
+        self._tsa_certificate: cryptography.x509.Certificate | None = tsa_certificate
+        self._intermediates: list[cryptography.x509.Certificate] = intermediates or []
+        self._roots: list[cryptography.x509.Certificate] = roots or []
+        self._nonce: int | None = nonce
+        self._common_name: str | None = common_name
+
+    def policy_id(self, policy_oid: cryptography.x509.ObjectIdentifier) -> VerifyBuilder:
+        """Set the nonce."""
+        if not isinstance(policy_oid, cryptography.x509.ObjectIdentifier):
+            msg = "The policy oid type is not valid"
+            raise ValueError(msg)
+        if self._policy_id is not None:
+            msg = "The policy id can be set only once"
+            raise ValueError(msg)
+        return VerifyBuilder(
+            policy_id=policy_oid,
+            tsa_certificate=self._tsa_certificate,
+            intermediates=self._intermediates,
+            roots=self._roots,
+            nonce=self._nonce,
+            common_name=self._common_name,
+        )
 
-def create_verify_opts(
-    tsp_request: TimeStampRequest,
-    tsa_certificate: cryptography.x509.Certificate | None,
-    common_name: str,
-    root_certificates: list[cryptography.x509.Certificate] | None = None,
-    intermediates: list[cryptography.x509.Certificate] | None = None,
-) -> VerifyOpts:
-    if intermediates is None:
-        intermediates = []
+    def tsa_certificate(self, certificate: cryptography.x509.Certificate) -> VerifyBuilder:
+        """Set the TSA certificate."""
+        if not isinstance(certificate, cryptography.x509.Certificate):
+            msg = "The certificate type is not valid"
+            raise ValueError(msg)
```

Comment:
Do we need these type checks? (here and in the other methods)

Reply:
Albeit a bit verbose, I think they make sense to ease the API usage.
However, I'm not opposed to deleting them if you think they are a bit redundant.

Reply:
IMO they're redundant -- if people don't use the type hints that's on them, and they'll get a type error soon enough anyways 🙂